onnxruntime/core/providers/cuda/tensor/tile_impl.cu

```diff
@@ -7,140 +7,178 @@
-    const size_t N) {
+__global__ void _TileMemcpyKernel(const T* input_data, T* output_data, const CUDA_LONG N, const size_t repeats) {
   CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
-  auto input_index = id % num_input_elements;
-  output_data[id] = input_data[input_index];
+  T input_val = input_data[id];
+  for (size_t i = 0; i < repeats; ++i) {
+    output_data[id] = input_val;
+    id += N;
+  }
 }
 
 template <typename T>
-void TileMemcpyImpl(
-    cudaStream_t stream,
-    const T* input_data,
-    const size_t num_input_elements,
-    T* output_data,
-    const size_t num_output_elements) {
-  int blocksPerGrid = (int)(ceil(static_cast<float>(num_output_elements) / GridDim::maxThreadsPerBlock));
-  _TileMemcpyKernel<<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
-      input_data, num_input_elements, output_data, (CUDA_LONG)num_output_elements);
+void TileMemcpyImpl(cudaStream_t stream, const T* input_data, T* output_data, const size_t num_input_elements,
+                    const size_t repeats) {
+  using Vec4T = aligned_vector<T, 4>;
+  using Vec2T = aligned_vector<T, 2>;
+  constexpr int vec4_alignment = std::alignment_of<Vec4T>::value;
+  constexpr int vec2_alignment = std::alignment_of<Vec2T>::value;
+  uint64_t address_input = reinterpret_cast<uint64_t>(input_data);
+  uint64_t address_output = reinterpret_cast<uint64_t>(output_data);
+  size_t N = num_input_elements;
+  if (num_input_elements % 4 == 0 && address_input % vec4_alignment == 0 && address_output % vec4_alignment == 0) {
+    N = num_input_elements / 4;
+    int blocksPerGrid = static_cast<int>(CeilDiv(N, num_threads_per_block));
+    _TileMemcpyKernel<<<blocksPerGrid, num_threads_per_block, 0, stream>>>(reinterpret_cast<const Vec4T*>(input_data),
+                                                                           reinterpret_cast<Vec4T*>(output_data),
+                                                                           static_cast<CUDA_LONG>(N), repeats);
+    return;
+  } else if (num_input_elements % 2 == 0 && address_input % vec2_alignment == 0 &&
+             address_output % vec2_alignment == 0) {
+    N = num_input_elements / 2;
+    int blocksPerGrid = static_cast<int>(CeilDiv(N, num_threads_per_block));
+    _TileMemcpyKernel<<<blocksPerGrid, num_threads_per_block, 0, stream>>>(reinterpret_cast<const Vec2T*>(input_data),
+                                                                           reinterpret_cast<Vec2T*>(output_data),
+                                                                           static_cast<CUDA_LONG>(N), repeats);
+    return;
+  }
+
+  int blocksPerGrid = static_cast<int>(CeilDiv(N, num_threads_per_block));
+  _TileMemcpyKernel<<<blocksPerGrid, num_threads_per_block, 0, stream>>>(input_data, output_data,
+                                                                         static_cast<CUDA_LONG>(N), repeats);
 }
 
+// Input size is [n, c], output size is [n * batch_repeats, c * repeats].
```

Comment:
Nit: Can we see data can effectively be viewed as [batch, data].... [n, c] seems like it only supports 2D input (atleast for me)